### PR TITLE
feat: Open Gov for Development and Altair Runtimes

### PR DESCRIPTION
**Pull request by wischli**
# Description

* Adds OpenGov for Development and Altair Runtimes which coexists with current Gov structure
* Altair config based on https://gov.centrifuge.io/t/rfc-altair-migration-to-opengov/6011
  * More thorough [Google Doc for Altair tracks and parametrization](https://docs.google.com/document/d/1asoDHtIT3bhSnwfQir3SZ7NN-5D1i02s01clh9lEd0k)
* Development config based on Centrifuge Chain configuration: https://gov.centrifuge.io/t/rfc-centrifuge-migration-to-opengov/5984
  * Time scale entirely mapped to 1 minute: 
```diff
// centrifuge runtime
x_period: 2 * MINUTES,
- y_period: 3 * HOURS,
- z_period: 4 * DAYS,

// development runtime
x_period: 1 * MINUTES,
+ y_period: 1 * MINUTES,
+ z_period: 1 * MINUTES,
```
  * More thorough [Google Doc for Centrifuge tracks and parametrization](https://docs.google.com/document/d/1aPH1sTfAttD3K0f4AoOF2iDb3nH1GHlx5TfIP4Xi4SM)

## Changes and Descriptions

* Adds `origins` module to `runtime_common` necessary for gov origin declaration which is the same across all runtimes
* Adds necessary OpenGov pallets to Development and Altair runtimes (Centrifuge will follow in ~1 month)
* Initializes Technical Committee for `WhitelistedCaller` track via migration based on [CP86](https://gov.centrifuge.io/t/cp86-opengov-technical-committee-mandate-request/6085)
  * Requires `pallet_membership` for membership handling within the collective and 
  * `pallet_collective::Instance2` for membership handling via root as well origin derivation for the `WhitelistedCaller` track

## TODO

All done 14.05.2024

- [x] OpenGov Development Runtime
- [x] OpenGov Altair Runtime
- [x] Revert OpenGov addition to Centrifuge Runtime
- [x] Write migration to initialize Technical Committee

## Altair migrations

```
[2024-05-14T13:00:39Z INFO  remote-ext] initialized state externalities with storage root 0xa907b41b327bb68ffcdd9902a0b5cffe8bccae5a2ccc7b7cab8fde43bc18c272 and state_version V0
[2024-05-14T13:00:40Z INFO  try-runtime::cli] Original runtime [Name: RuntimeString::Owned("altair")] [Version: 1034] [Code hash: 0xbdc0...5150]
[2024-05-14T13:00:40Z INFO  try-runtime::cli] New runtime      [Name: RuntimeString::Owned("altair")] [Version: 1035] [Code hash: 0x7f0d...4cf8]
[2024-05-14T13:00:40Z INFO  try_runtime_core::commands::on_runtime_upgrade] 🔬 Running TryRuntime_on_runtime_upgrade with checks: PreAndPost
[2024-05-14T13:00:41Z INFO  runtime_common::migrations::increase_storage_version] BumpStorageVersion: Increasing storage version of "OraclePriceFeed" from StorageVersion(0) to StorageVersion(1)
[2024-05-14T13:00:41Z INFO  runtime_common::migrations::increase_storage_version] BumpStorageVersion: Increasing storage version of "OraclePriceCollection" from StorageVersion(0) to StorageVersion(1)
[2024-05-14T13:00:41Z INFO  runtime_common::migrations::increase_storage_version] BumpStorageVersion: Increasing storage version of "OrderBook" from StorageVersion(0) to StorageVersion(1)
[2024-05-14T13:00:41Z INFO  runtime_common::migrations::increase_storage_version] BumpStorageVersion: Increasing storage version of "ForeignInvestments" from StorageVersion(0) to StorageVersion(1)
[2024-05-14T13:00:41Z INFO  runtime::collator-selection] Sorted 3 Invulnerables, upgraded storage to version 1
[2024-05-14T13:00:41Z INFO  runtime_common::migrations::technical_comittee] InitTechnicalCommittee: Pre checks done
[2024-05-14T13:00:41Z INFO  runtime_common::migrations::technical_comittee] InitTechnicalCommittee: Setting up members
[2024-05-14T13:00:41Z INFO  runtime_common::migrations::technical_comittee] InitTechnicalCommittee: Migration done
[2024-05-14T13:00:41Z INFO  runtime_common::migrations::technical_comittee] InitTechnicalCommittee: Post checks done
[2024-05-14T13:00:41Z INFO  runtime_common::migrations::increase_storage_version] BumpStorageVersion: Increasing storage version of "Referenda" from StorageVersion(0) to StorageVersion(1)
[2024-05-14T13:00:41Z INFO  runtime_common::migrations::increase_storage_version] BumpStorageVersion: Increasing storage version of "TechnicalCommittee" from StorageVersion(0) to StorageVersion(4)
[2024-05-14T13:00:41Z INFO  runtime_common::migrations::increase_storage_version] BumpStorageVersion: Increasing storage version of "TechnicalCommitteeMembership" from StorageVersion(0) to StorageVersion(4)
[2024-05-14T13:00:41Z INFO  runtime::frame-support] ⚠️ Ethereum declares internal migrations (which *might* execute). On-chain `StorageVersion(0)` vs current storage version `NoStorageVersionSet`
[2024-05-14T13:00:41Z INFO  try_runtime_core::commands::on_runtime_upgrade] 🔬 TryRuntime_on_runtime_upgrade succeeded! Running it again without checks for weight measurements.
[2024-05-14T13:00:41Z INFO  runtime_common::migrations::increase_storage_version] BumpStorageVersion: Increasing storage version of "OraclePriceFeed" from StorageVersion(0) to StorageVersion(1)
[2024-05-14T13:00:41Z INFO  runtime_common::migrations::increase_storage_version] BumpStorageVersion: Increasing storage version of "OraclePriceCollection" from StorageVersion(0) to StorageVersion(1)
[2024-05-14T13:00:41Z INFO  runtime_common::migrations::increase_storage_version] BumpStorageVersion: Increasing storage version of "OrderBook" from StorageVersion(0) to StorageVersion(1)
[2024-05-14T13:00:41Z INFO  runtime_common::migrations::increase_storage_version] BumpStorageVersion: Increasing storage version of "ForeignInvestments" from StorageVersion(0) to StorageVersion(1)
[2024-05-14T13:00:41Z INFO  runtime::collator-selection] Sorted 3 Invulnerables, upgraded storage to version 1
[2024-05-14T13:00:41Z INFO  runtime_common::migrations::technical_comittee] InitTechnicalCommittee: Setting up members
[2024-05-14T13:00:41Z INFO  runtime_common::migrations::technical_comittee] InitTechnicalCommittee: Migration done
[2024-05-14T13:00:41Z INFO  runtime_common::migrations::increase_storage_version] BumpStorageVersion: Increasing storage version of "Referenda" from StorageVersion(0) to StorageVersion(1)
[2024-05-14T13:00:41Z INFO  runtime_common::migrations::increase_storage_version] BumpStorageVersion: Increasing storage version of "TechnicalCommittee" from StorageVersion(0) to StorageVersion(4)
[2024-05-14T13:00:41Z INFO  runtime_common::migrations::increase_storage_version] BumpStorageVersion: Increasing storage version of "TechnicalCommitteeMembership" from StorageVersion(0) to StorageVersion(4)
[2024-05-14T13:00:41Z INFO  runtime::frame-support] ⚠️ Ethereum declares internal migrations (which *might* execute). On-chain `StorageVersion(0)` vs current storage version `NoStorageVersionSet`
[2024-05-14T13:00:41Z INFO  try-runtime::cli] PoV size (zstd-compressed compact proof): 1.4 KB. For parachains, it's your responsibility to verify that a PoV of this size fits within any relaychain constraints.
[2024-05-14T13:00:41Z INFO  try-runtime::cli] Consumed ref_time: 0.001691103s (0.34% of max 0.5s)
[2024-05-14T13:00:41Z INFO  try-runtime::cli] ✅ TryRuntime_on_runtime_upgrade executed without errors or weight safety warnings. Please note this does not guarantee a successful runtime upgrade. Always test your runtime upgrade with recent state, and ensure that the weight usage of your migrations will not drastically differ between testing and actual on-chain execution.
```

# Checklist:

- [x] I have added Rust doc comments to structs, enums, traits and functions
- [x] I have made corresponding changes to the documentation
- [x] I have performed a self-review of my code
- [x] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

File: libs/primitives/src/lib.rs
Comment (lemunozm) [inline libs/primitives/src/lib.rs:58]: Could they go in `runtime/common/src/lib.rs` under `instances`?

File: runtime/altair/Cargo.toml
Comment (lemunozm) [inline runtime/altair/Cargo.toml:381]: Curious about these changes. Are you using some code formatter for `.toml` files other than `taplo fmt`?

File: runtime/altair/src/governance/mod.rs
Comment (lemunozm) [inline runtime/altair/src/governance/mod.rs:50]: NIT and open to discussion:

I feel that splitting the runtime config implementations into different files ends up reducing the readability and maintainability because both splitting files end up with mutual self-references using types from the other, and sometimes it is very difficult to find the implementation for a pallet if these files grow

I'm a big fan of splitting things, but I would leave all `impl xyz::Config for Runtime` parts in the same file as an exception.

File: runtime/centrifuge/src/governance/tracks.rs
Comment (lemunozm) [inline runtime/centrifuge/src/governance/tracks.rs:161]: Could we abstract these files in `runtime-common`? Maybe we can pass some config type with the changed values in each runtime

**Review by lemunozm — COMMENTED** [review]
Thanks so much for dive into this topic @wischli!

I left just some minor comments in a fast review

File: runtime/altair/Cargo.toml
Comment (wischli) [reply runtime/altair/Cargo.toml:381]: I just used `taplo fmt` to fix the CI complaint. Also surprised it resulted in such a large diff.

File: runtime/centrifuge/src/governance/tracks.rs
Comment (wischli) [reply runtime/centrifuge/src/governance/tracks.rs:161]: We could move three scopes into `runtime-common`:
1. `impl pallet_referenda::TracksInfo<Balance, BlockNumber> for TracksInfo` block
2. `fn percent`
3. `fn ppm`

The rest is completely custom to each runtime and I would imagine readability to suffer just for moving a few more lines into common.

Comment (lemunozm) [reply runtime/centrifuge/src/governance/tracks.rs:161]: Makes sense!

File: runtime/altair/src/governance/mod.rs
Comment (wischli) [reply runtime/altair/src/governance/mod.rs:50]: I already knew you would complain about this. Absolutely fine for me to move it into the same place. I just preferred having all separate as long as this PR is a draft and not open for reviews :)

File: runtime/altair/Cargo.toml
Comment (lemunozm) [reply runtime/altair/Cargo.toml:381]: That weird. It also removes blank lines and increases the indentation. My `taplo` does not do that 😆. Which version are you using? Could you have some extra `.toml` plugin in your IDE?

File: runtime/altair/src/governance/mod.rs
Comment (lemunozm) [reply runtime/altair/src/governance/mod.rs:50]: Sorry for my early review! 🙏🏻 Just wanna to highlight the topic

File: runtime/centrifuge/src/governance/tracks.rs
Comment (wischli) [reply runtime/centrifuge/src/governance/tracks.rs:161]: Unfortunately, we cannot move the large `impl pallet_referenda::TracksInfo<Balance, BlockNumber> for TracksInfo` block to common because it requires two inputs: `RuntimeOrigin` from the runtime and the `TRACKS_DATA` array which is custom to each runtime.

I tried to get around making both these parameters generic but the `pallet_referenda::impl_tracksinfo_get!(TracksInfo, Balance, BlockNumber)` macro then acts as a blocker 😞

File: libs/primitives/src/lib.rs
Comment (wischli) [reply libs/primitives/src/lib.rs:58]: I am not sure whether this makes sense. I added these two instances here because the existing `CouncilCollective` is also defined in `primitives::types`. We cannot move it out because the `AllOfCouncil`, `HalfOfCouncil` and similar types of `primitives::types` are dependent on `CouncilCollective` and `primitives` should not have `common` as dependency.

I can imagine that we might eventually define similar types (`AllOfTechnical...` etc.) for the `TechnicalCollective` which should probably stay in `primitives::types`?

File: runtime/centrifuge/src/governance/tracks.rs
Comment (lemunozm) [reply runtime/centrifuge/src/governance/tracks.rs:161]: Then something like: 

```rust
pallet_referenda::impl_tracksinfo_get!(TracksInfo<Origin, TRACKS_DATA>, Balance, BlockNumber);
```

called in each runtime doesn't work, right?

No worries! Thanks for trying this! hopefully if we condense all runtimes into one in the future using a macro for the different configuration we can simplify this

Comment (wischli) [reply runtime/centrifuge/src/governance/tracks.rs]: This is the best I could do: [a26245a](https://github.com/centrifuge/centrifuge-chain/pull/1828/commits/a26245a88cb9faadc3c0441092cffd4a89411859) + [eebb23f](https://github.com/centrifuge/centrifuge-chain/pull/1828/commits/eebb23f8a69b85043e0703a874f8900e4335ba53)

File: libs/primitives/src/lib.rs
Comment (lemunozm) [reply libs/primitives/src/lib.rs]: I would move the current `AllOfCouncil` and `CouncilCollective` to `runtime-common`. For me, primitives should not know about pallets or how runtimes are configured. WDYT?

Comment (wischli) [reply libs/primitives/src/lib.rs]: That's what I did in [f530ee559c6a0575826ec67780cbb426d7e54f16](https://github.com/centrifuge/centrifuge-chain/pull/1828/commits/f530ee559c6a0575826ec67780cbb426d7e54f16#diff-8930d436577b42479d8367c28afd439e7bb65a9f14ed350201685ba0602701b4). I apologize for the difficulty to follow this properly because of my continuous force rebases. I just wanted to have the Centrifuge OpenGov commit as clean as possible so I can cherry-pick it in a month or so and have as few conflicts as possible.

Comment (lemunozm) [reply libs/primitives/src/lib.rs]: Ah cool! I didn't see it. No worries! Thanks

File: runtime/altair/Cargo.toml
Comment (wischli) [reply runtime/altair/Cargo.toml]: I don't have any extra toml plugin and I execute `taplo fmt` from an external terminal `iTerm` anyways. I updated to the newest version of taplo `v0.7.0` hoping to result in a smaller diff but that unfortunately that didn't happen.

Comment (lemunozm) [reply runtime/altair/Cargo.toml]: I'm using `0.9.0` and I think CI too

Comment (wischli) [reply runtime/altair/Cargo.toml]: For some reason it's fixed now 🤷

File: runtime/development/src/lib.rs
Comment (wischli) [inline runtime/development/src/lib.rs:1366]: OpenGov introduces a `PoolAdmin` track just for pool related stuff. Of course, this includes pool creation. IMO it also makes sense to include the permissions and keystore `AdminOrigin`. I would be interested in your thoughts @mustermeiszer @hieronx. For Altair, it does not matter too much probably but in the end Altair and Centrifuge should have the same configuration apart from the track parameters (deposits, durations, voting thresholds).

The difference between the `Root` and `PoolAdmin` tracks is that the latter has much lower thresholds in terms of deposits, support and it is faster:

Root
Screenshot 2024-05-14 at 15 16 49

Pool Admin
Screenshot 2024-05-14 at 15 16 18

File: runtime/altair/src/lib.rs
Comment (wischli) [inline runtime/altair/src/lib.rs:1048]: OpenGov introduces a `PoolAdmin` track just for pool related stuff. Of course, this includes pool creation. IMO it also makes sense to include the permissions and keystore `AdminOrigin`. I would be interested in your thoughts @mustermeiszer @hieronx. For Altair, it does not matter too much probably but in the end Altair and Centrifuge should have the same configuration apart from the track parameters (deposits, durations, voting thresholds).

The difference between the `Root` and `PoolAdmin` tracks is that the latter has much lower thresholds in terms of deposits, support and it is faster:

Root
Screenshot 2024-05-14 at 15 16 49

Pool Admin
Screenshot 2024-05-14 at 15 16 18

File: runtime/altair/src/lib.rs
Comment (wischli) [inline runtime/altair/src/lib.rs:2107]: NOTE: Moved to top such that all dependencies are imported in the header. Having two separate places messed up IntelliJ's import helper because it seems to add imports to the location which is closer in terms of LoC.

File: runtime/altair/src/lib.rs
Comment (wischli) [inline runtime/altair/src/lib.rs:2772]: NOTE: We now have two benchmarks for the `pallet_collective` based on the `CouncilCollective` and `TechnicalCollective` instances.

File: runtime/altair/src/migrations.rs
Comment (wischli) [inline runtime/altair/src/migrations.rs:17]: NOTE: Missing 4 members at the time of writing. I'd like this PR to get merged rather sooner than later such that a subsequent PR to add the missing ones is preferred from my side. But no strong opinion.

File: runtime/altair/Cargo.toml
Comment (lemunozm) [inline runtime/altair/Cargo.toml:272]: Where is this crate used?

File: runtime/common/src/lib.rs
Comment (lemunozm) [inline runtime/common/src/lib.rs:814]: Super NIT. Maybe it fits better in cfg_utils/maths or similar (?)

(also applies to `to_ppm()`)

File: runtime/common/src/migrations/technical_comittee.rs
Comment (lemunozm) [inline runtime/common/src/migrations/technical_comittee.rs:49]: We probably want to apply this migration if the members are already there. i.e. a new member of the team should overwrite the member list with the new addition.

File: runtime/common/src/origins.rs
Comment (lemunozm) [inline runtime/common/src/origins.rs:29]: To understand this pallet correctly. The idea behind this is to add these variants to the composed `RuntimeOrigin` automatically once this pallet is added by the runtime, right?

Pretty smart!

File: runtime/development/src/migrations.rs
Comment (lemunozm) [inline runtime/development/src/migrations.rs:58]: Just curious, a pallet just added to the runtime, does not contain the latest version?

**Review by lemunozm — COMMENTED** [review]
Thanks for your amazing work on this @wischli 

Just one question maybe I miss, why centrifuge does not use this new open gov? Are we waiting to test it before in altair/dev?

File: runtime/common/src/migrations/technical_comittee.rs
Comment (wischli) [reply runtime/common/src/migrations/technical_comittee.rs:49]: We dont need the migration if there is a member because the membership pallet provides us with means of **adding** members via motions on which all TC members vote. Without the membership pallet, we would always have to fall back to the root track to **set** the TC members

File: runtime/development/src/migrations.rs
Comment (wischli) [reply runtime/development/src/migrations.rs:58]: Apparently! I have to admit that I did not dig deeper into this but `try-runtime` complained about it. It really surprised me!

File: runtime/common/src/migrations/technical_comittee.rs
Comment (lemunozm) [reply runtime/common/src/migrations/technical_comittee.rs:49]: Ahh ok, then great, thanks for answering!

File: runtime/common/src/lib.rs
Comment (wischli) [reply runtime/common/src/lib.rs]: Absolutely agree: [c7be7f9](https://github.com/centrifuge/centrifuge-chain/pull/1828/commits/c7be7f968abfb789ddb949affc9a6cce98a7ed81)

File: runtime/altair/Cargo.toml
Comment (wischli) [reply runtime/altair/Cargo.toml]: Great find! Turns out we have used this crate throughout all runtimes without needing it. Moreover, my commit duplicated a few `std` derivations for some crates.

* Cleanup of multiple deps in Altair: [9a4950a](https://github.com/centrifuge/centrifuge-chain/pull/1828/commits/9a4950a5095dac845053a980fa6484de68717bae)
* Removal of `getrandom` for all runtimes: [bc2ae52](https://github.com/centrifuge/centrifuge-chain/pull/1828/commits/bc2ae5273ac07eb93c38806fe0db0294601815d9)

File: runtime/altair/src/lib.rs
Comment (lemunozm) [reply runtime/altair/src/lib.rs:2107]: Yeah, it comes from removing the runtime-api feature. There were some `use` statements below I forgot to go up

File: runtime/altair/Cargo.toml
Comment (lemunozm) [reply runtime/altair/Cargo.toml]: Thanks!!

File: runtime/common/src/lib.rs
Comment (lemunozm) [reply runtime/common/src/lib.rs]: Thanks for tackle my Nit thing 🙌🏻

**Review by lemunozm — APPROVED** [review]
For me everything looks great but I would like another pair of eyes, at least in the configuration part of the governance.

**Comment by lemunozm** [comment]
@wischli Do you expect to merge this before or after 1.7.2?

File: runtime/integration-tests/src/generic/utils/democracy.rs
Comment (lemunozm) [inline runtime/integration-tests/src/generic/utils/democracy.rs:3]: With this PR, do we no longer need the old democracy? If not, I think this file can be fully removed. Just the gateway tests are using them, but they can done though Root instead.

**Review by lemunozm — DISMISSED** [review]
Thanks for leading this change! Organized and easy to follow. LGTM!

**Comment by wischli** [comment]
Just FYI, I have ensured OpenGov works properly after rebasing to main as the initial checks were done with SDK v1.3.0.

File: runtime/altair/src/lib.rs
Comment (wischli) [inline runtime/altair/src/lib.rs:176]: FYI: Bumping to 1200 here and removing outdated migrations so I can use this as RC for Altair.

**Review by mustermeiszer — APPROVED** [review]
Nothing to complain from my side. Thanks for the very detailed and great PR documentation and insights into the choosing of the values. Lets see how this works out in the testnets.

File: runtime/centrifuge/src/migrations.rs
Comment (lemunozm) [inline runtime/centrifuge/src/migrations.rs:15]: Ok, we need to be sure we use the 1103 as RU to apply the previous migrations 👍🏻

**Pull request merged**
